### PR TITLE
Refactor noise model

### PR DESCRIPTION
**Pull request by liwt31**
- Refactor noise model. 
    - Noise configurations are represented by tuples of description, condition and Kraus operators. 
    - Reatout error is treated separately. Compared to other noises, currently, the configuration of readout error is pretty simple. Probably it should be refactored to enable more functionalities such as different readout error for different qubits in the future. (edit: this is not  true)
    - Add `add_noise_by_condition`. 
    - Composing noise channels is disabled in `noisemodel.py`
    - Make the noise model tests more strict
- Enhance `gate_count`. Now support string inputs.
- Fix bug in the scipy interface to avoid complex to float warnings by numpy.

### CODE REVIEW

**Comment by refraction-ray** [comment]
for point 2, I think we already support different readout support on different qubit?

File: tensorcircuit/noisemodel.py [hunk 1]
```diff
@@ -57,40 +59,79 @@ def add_noise(
         :param qubit: the list of noisy qubit, defaults to None, indicating applying the noise channel on all qubits
         :type qubit: Optional[Sequence[Any]], optional
         """
-        if gate_name != "readout":
-            gate_name = AbstractCircuit.standardize_gate(gate_name)
+        if gate_name == "readout":
+            # probably need to refactor readout error
+            assert qubit is None
```
Comment (refraction-ray) [inline tensorcircuit/noisemodel.py:64]: we should support different readout on different qubit, this is a level-1 requirements

File: tests/test_noisemodel.py [hunk 2]
```diff
@@ -132,20 +138,20 @@ def test_general_noisemodel(backend):
     value1 = sample_expectation_ps_noisfy(c, x=[0, 1], noise_conf=noise_conf, nmc=10000)
     value2 = c.sample_expectation_ps(x=[0, 1], noise_conf=noise_conf, nmc=10000)
     value3 = dmc.sample_expectation_ps(x=[0, 1], noise_conf=noise_conf)
-    np.testing.assert_allclose(value1, value2, atol=1e-1)
-    np.testing.assert_allclose(value3, value2, atol=1e-1)
+    np.testing.assert_allclose(value1, value2, atol=1e-2)
+    np.testing.assert_allclose(value3, value2, atol=1e-2)
```
Comment (refraction-ray) [inline tests/test_noisemodel.py:142]: a stricter checker for sample expectation may fail the test in rare case, unless the strict criteria is still rather flexible

**Comment by liwt31** [comment]
> for point 2, I think we already support different readout support on different qubit?

My bad. It is indeed already supported. The refactoring does not break the feature as it's included in the test suite

File: tests/test_noisemodel.py (hunk 2 above)
Comment (liwt31) [reply tests/test_noisemodel.py:142]: Then I think it's better to increase the number of runs in MC simulations. `1e-1` is too lose IMO

File: tensorcircuit/noisemodel.py (hunk 1 above)
Comment (refraction-ray) [reply tensorcircuit/noisemodel.py:64]: should be auto satisfied

File: tensorcircuit/noisemodel.py [hunk 3]
```diff
@@ -57,40 +59,79 @@ def add_noise(
...
+
+        :param c: the circuit to be counted
+        :type c: Circuit
+        :return: the count
+        :rtype: int
+        """
+        count = 0
+        for d in c.to_qir():
+            # pylint: disable-next=unused-variable
+            for _, condition, krauslist in self.nc:
```
Comment (refraction-ray) [inline tensorcircuit/noisemodel.py:131]: why we use krauslist instead of _ here?

Comment (liwt31) [reply tensorcircuit/noisemodel.py:131]: Didn't notice. I thought it's complaining about the first `_`. Will fix it in the next commit.

File: tests/test_noisemodel.py (hunk 2 above)
Comment (refraction-ray) [reply tests/test_noisemodel.py:142]: you can check the value now, I think the criteria can be one order larger than the average difference in trials

**Comment by refraction-ray** [comment]
other parts look good to me, only the test criteria need double check

File: tests/test_noisemodel.py (hunk 2 above)
Comment (liwt31) [reply tests/test_noisemodel.py:142]: `nmc=100000` reduces the std deviation to around 1e-3

Comment (refraction-ray) [reply tests/test_noisemodel.py:142]: then I think it is ok

**Review by refraction-ray — APPROVED** [review]
LGTM

**Pull request merged**